File: code/client/munkilib/iconutils.py
# encoding: utf-8
#
# Copyright 2010-2023 Greg Neagle.
#
# Licensed under the Apache License, Version 2.0 (the 'License');
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an 'AS IS' BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
iconutils

Created by Greg Neagle on 2014-05-15.

Functions to work with product images ('icons') for Managed Software Center
"""
from __future__ import absolute_import, print_function

import glob
import os
import shutil
import subprocess
import tempfile

# PyLint cannot properly find names inside Cocoa libraries, so issues bogus
# No name 'Foo' in module 'Bar' warnings. Disable them.
# pylint: disable=E0611
from Foundation import NSURL
from Quartz import (CGImageSourceCreateWithURL, CGImageSourceCreateImageAtIndex,
                    CGImageDestinationCreateWithURL, CGImageDestinationAddImage,
                    CGImageDestinationFinalize,
                    CGImageSourceGetCount, CGImageSourceCopyPropertiesAtIndex,
                    kCGImagePropertyDPIHeight, kCGImagePropertyPixelHeight)
# pylint: enable=E0611

from . import display
from .wrappers import readPlist, PlistReadError


# we use lots of camelCase-style names. Deal with it.
# pylint: disable=C0103


def convertIconToPNG(icon_path, destination_path,
                     desired_pixel_height=350, desired_dpi=72):
    '''Converts an icns file to a png file, choosing the representation
    closest to (but >= if possible) the desired_pixel_height.
    Returns True if successful, False otherwise'''
    icns_url = NSURL.fileURLWithPath_(icon_path)
    png_url = NSURL.fileURLWithPath_(destination_path)

    image_source = CGImageSourceCreateWithURL(icns_url, None)
    if not image_source:
        return False
    number_of_images = CGImageSourceGetCount(image_source)
    if number_of_images == 0:
        return False

    selected_index = 0
    candidate = {}
    # iterate through the individual icon sizes to find the "best" one
    for index in range(number_of_images):
        try:
            properties = CGImageSourceCopyPropertiesAtIndex(
                image_source, index, None)
            # perform not empty check for properties to prevent crash as CGImageSourceCopyPropertiesAtIndex sometimes fails in 10.15.4 and above
            if not properties:
                return False
            dpi = int(properties.get(kCGImagePropertyDPIHeight, 0))
            height = int(properties.get(kCGImagePropertyPixelHeight, 0))
            if (not candidate or
                    (height < desired_pixel_height and
                     height > candidate['height']) or
                    (height >= desired_pixel_height and
                     height < candidate['height']) or
                    (height == candidate['height'] and dpi == desired_dpi)):
                candidate = {'index': index, 'dpi': dpi, 'height': height}
                selected_index = index
        except ValueError:
            pass

    image = CGImageSourceCreateImageAtIndex(image_source, selected_index, None)
    image_dest = CGImageDestinationCreateWithURL(png_url, 'public.png', 1, None)
    CGImageDestinationAddImage(image_dest, image, None)
    return CGImageDestinationFinalize(image_dest)


def findIconForApp(app_path):
    '''Finds the icon file for app_path. Returns a path or None.'''
    if not os.path.exists(app_path):
        return None
    try:
        info = readPlist(
            os.path.join(app_path, u'Contents/Info.plist'))
    except PlistReadError:
        return None
    app_name = os.path.basename(app_path)
    icon_filename = info.get('CFBundleIconFile', app_name)
    icon_path = os.path.join(app_path, u'Contents/Resources', icon_filename)
    if not os.path.splitext(icon_path)[1]:
        # no file extension, so add '.icns'
        icon_path += '.icns'
    if os.path.exists(icon_path):
        return icon_path
    return None


def extractAppBitsFromPkgArchive(archive_path, target_dir):
    '''Extracts application Info.plist and .icns files into target_dir
       from a package archive file. Returns the result code of the
       pax extract operation.'''
    result = -999
    if os.path.exists(archive_path):
        original_dir = os.getcwd()
        os.chdir(target_dir)
        cmd = ['/bin/pax', '-rzf', archive_path,
               '*.app/Contents/Info.plist',
               '*.app/Contents/Resources/*.icns']
        result = subprocess.call(cmd)
        os.chdir(original_dir)
    return result


def extractAppIconsFromFlatPkg(pkg_path):
    '''Extracts application icons from a flat package.
       Returns a list of paths to icns files.'''
    cmd = ['/usr/sbin/pkgutil', '--bom', pkg_path]
    proc = subprocess.Popen(cmd, shell=False, bufsize=-1,
                            stdin=subprocess.PIPE, stdout=subprocess.PIPE,
                            stderr=subprocess.STDOUT)
    output = proc.communicate()[0].decode('UTF-8')
    if proc.returncode:
        display.display_error(u'Could not get bom files from %s', pkg_path)
        return []
    bomfilepaths = output.splitlines()
    pkg_dict = {}
    for bomfile in bomfilepaths:
        # bomfile path is of the form:
        # /tmp/FlashPlayer.pkg.boms.2Rxa1z/AdobeFlashPlayerComponent.pkg/Bom
        pkgname = os.path.basename(os.path.dirname(bomfile))
        if not pkgname.endswith(u'.pkg'):
            # no subpackages; this is a component pkg
            pkgname = ''
        cmd = ['/usr/bin/lsbom', '-s', bomfile]
        proc = subprocess.Popen(cmd, shell=False, bufsize=-1,
                                stdin=subprocess.PIPE, stdout=subprocess.PIPE,
                                stderr=subprocess.STDOUT)
        output = proc.communicate()[0].decode('UTF-8')
        if proc.returncode:
            display.display_error(u'Could not lsbom %s', bomfile)
        # record paths to all app Info.plist files
        pkg_dict[pkgname] = [
            os.path.normpath(line)
            for line in output.splitlines()
            if line.endswith(u'.app/Contents/Info.plist')]
        if not pkg_dict[pkgname]:
            # remove empty lists
            del pkg_dict[pkgname]
    if not pkg_dict:
        return []
    icon_paths = []
    pkgtmp = os.path.join(tempfile.mkdtemp(dir=u'/tmp'), u'pkg')
    exporttmp = tempfile.mkdtemp(dir='/tmp')
    cmd = ['/usr/sbin/pkgutil', '--expand', pkg_path, pkgtmp]
    result = subprocess.call(cmd)
    if result == 0:
        for pkg in pkg_dict:
            archive_path = os.path.join(pkgtmp, pkg, u'Payload')
            err = extractAppBitsFromPkgArchive(archive_path, exporttmp)
            if err == 0:
                for info_path in pkg_dict[pkg]:
                    full_path = os.path.join(exporttmp, info_path)
                    # convert path to Info.plist to path to app
                    app_path = os.path.dirname(os.path.dirname(full_path))
                    icon_path = findIconForApp(app_path)
                    if icon_path:
                        icon_paths.append(icon_path)
            else:
                display.display_error(
                    u'pax could not read files from %s', archive_path)
                return []
    else:
        display.display_error(u'Could not expand %s', pkg_path)
    # clean up our expanded flat package; we no longer need it
    shutil.rmtree(pkgtmp)
    return icon_paths


def findInfoPlistPathsInBundlePkg(pkg_path, repo=None):
    '''Returns a dict with pkg paths as keys and filename lists
    as values'''
    pkg_dict = {}
    if repo:
        repo_bomfile = repo.join(pkg_path, u'Contents/Archive.bom')
        handle = repo.open(repo_bomfile, 'r')
        bomfile = handle.local_path
    else:
        bomfile = os.path.join(pkg_path, u'Contents/Archive.bom')
    if os.path.exists(bomfile):
        info_paths = getAppInfoPathsFromBOM(bomfile)
        if info_paths:
            pkg_dict[pkg_path] = info_paths
    else:
        # mpkg or dist pkg; look for component pkgs within
        pkg_dict = {}
        pkgs = []
        if repo:
            pkg_contents_dir = repo.join(pkg_path, u'Contents')
            if repo.isdir(pkg_contents_dir):
                pkgs = repo.glob(
                    pkg_contents_dir, '*.pkg', '*/*.pkg', '*/*/*.pkg',
                    '*.mpkg', '*/*.mpkg', '*/*/*.mpkg')
        else:
            pkg_contents_dir = os.path.join(pkg_path, u'Contents')
            if os.path.isdir(pkg_contents_dir):
                original_dir = os.getcwd()
                os.chdir(pkg_contents_dir)
                pkgs = (glob.glob('*.pkg') + glob.glob('*/*.pkg')
                        + glob.glob('*/*/*.pkg') + glob.glob('*.mpkg') +
                        glob.glob('*/*.mpkg') + glob.glob('*/*/*.mpkg'))
                os.chdir(original_dir)
        for pkg in pkgs:
            full_path = os.path.join(pkg_contents_dir, pkg)
            pkg_dict.update(findInfoPlistPathsInBundlePkg(full_path))
    return pkg_dict


def extractAppIconsFromBundlePkg(pkg_path, repo=None):
    '''Returns a list of paths for application icons found
    inside the bundle pkg at pkg_path'''
    pkg_dict = findInfoPlistPathsInBundlePkg(pkg_path, repo)
    icon_paths = []
    exporttmp = tempfile.mkdtemp(dir='/tmp')
    for pkg in pkg_dict:
        handle = None
        if repo:
            repo_archive_path = repo.join(pkg, u'Contents/Archive.pax.gz')
            handle = repo.open(repo_archive_path, 'r')
            archive_path = handle.local_path
        else:
            archive_path = os.path.join(pkg, u'Contents/Archive.pax.gz')
        err = extractAppBitsFromPkgArchive(archive_path, exporttmp)
        if err == 0:
            for info_path in pkg_dict[pkg]:
                full_path = os.path.normpath(os.path.join(exporttmp, info_path))
                app_path = os.path.dirname(os.path.dirname(full_path))
                icon_path = findIconForApp(app_path)
                if icon_path:
                    icon_paths.append(icon_path)
    return icon_paths


def getAppInfoPathsFromBOM(bomfile):
    '''Returns a list of paths to application Info.plists'''
    if os.path.exists(bomfile):
        cmd = ['/usr/bin/lsbom', '-s', bomfile]
        proc = subprocess.Popen(cmd, shell=False, bufsize=-1,
                                stdin=subprocess.PIPE, stdout=subprocess.PIPE,
                                stderr=subprocess.STDOUT)
        output = proc.communicate()[0].decode('UTF-8')
        return [line for line in output.splitlines()
                if line.endswith('.app/Contents/Info.plist')]
    return []


if __name__ == '__main__':
    print('This is a library of support tools for the Munki Suite.')

File: code/cli/munki/shared/utils/iconutils.swift
//
//  iconutils.swift
//  munki
//
//  Created by Greg Neagle on 7/11/24.
//
//  Copyright 2024-2026 The Munki Project.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//       https://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

import Foundation
import ImageIO

/// Converts an icns file to a png file.
/// desiredHeight should be one of the "native" icns sizes, like
/// 512, 256, 128, 48, 32, or 16
/// desiredDPI should be either 72 or 144
/// Returns true if successful, false otherwise
func convertIconToPNG(iconPath: String,
                      destinationPath: String,
                      desiredHeight: Int = 512,
                      desiredDPI: Int = 72) -> Bool
{
    typealias CandidateType = (height: Int, dpi: Int, index: Int)

    let iconURL = URL(fileURLWithPath: iconPath) as CFURL
    let pngURL = URL(fileURLWithPath: destinationPath) as CFURL
    guard let imageSource = CGImageSourceCreateWithURL(iconURL, nil) else { return false }
    let imageCount = CGImageSourceGetCount(imageSource)
    if imageCount == 0 {
        return false
    }
    var selectedIndex = -1
    var candidates = [CandidateType]()
    for index in 0 ..< imageCount {
        guard let properties = CGImageSourceCopyPropertiesAtIndex(imageSource, index, nil) as? PlistDict else {
            return false
        }
        let dpi = properties[kCGImagePropertyDPIHeight as String] as? Int ?? 0
        let height = properties[kCGImagePropertyPixelHeight as String] as? Int ?? 0
        if height == desiredHeight, dpi == desiredDPI {
            // we found one with our desired height and dpi
            selectedIndex = index
            break
        }
        candidates.append((height: height, dpi: dpi, index: index))
    }
    if selectedIndex == -1 {
        // didn't find an exact match, so look through the candidates and pick the best match
        for candidate in candidates {
            // is there a candidate with the desiredHeight, but wrong dpi? Use that
            if candidate.height == desiredHeight {
                selectedIndex = candidate.index
                break
            }
        }
    }
    if selectedIndex == -1 {
        // OK, now look for one with a height bigger than desired, but right dpi
        candidates.sort { $0 < $1 } // sort ascending, lowest height images first
        for candidate in candidates {
            // select the first candidate bigger than the desiredHeight with matching dpi
            if candidate.height > desiredHeight, candidate.dpi == desiredDPI {
                selectedIndex = candidate.index
                break
            }
        }
    }
    if selectedIndex == -1 {
        // still didn't find what we looked for, try again, and ignore DPI
        for candidate in candidates {
            if candidate.height > desiredHeight {
                selectedIndex = candidate.index
                break
            }
        }
    }
    if selectedIndex == -1 {
        // just grab the largest image
        candidates.sort { $0 > $1 } // sort descending, highest height images first
        selectedIndex = candidates[0].index
    }

    if let image = CGImageSourceCreateImageAtIndex(imageSource, selectedIndex, nil),
       let imageDestination = CGImageDestinationCreateWithURL(
           pngURL, "public.png" as CFString, 1, nil
       )
    {
        CGImageDestinationAddImage(imageDestination, image, nil)
        return CGImageDestinationFinalize(imageDestination)
    }

    return false
}

/// Finds the icon file for app_path. Returns a path or nil
func findIconForApp(_ appPath: String) -> String? {
    guard pathIsDirectory(appPath) else { return nil }
    let infoPlistPath = (appPath as NSString).appendingPathComponent("Contents/Info.plist")
    guard let info = try? readPlist(fromFile: infoPlistPath) as? PlistDict else { return nil }
    let appName = (appPath as NSString).lastPathComponent
    var iconFilename = info["CFBundleIconName"] as? String ?? info["CFBundleIconFile"] as? String ?? appName
    if (iconFilename as NSString).pathExtension.isEmpty {
        iconFilename += ".icns"
    }
    let iconPath = (appPath as NSString).appendingPathComponent(
        "Contents/Resources/\(iconFilename)")
    if pathIsRegularFile(iconPath) || pathIsSymlink(iconPath) {
        return iconPath
    }
    return nil
}

/// Extracts application Info.plist and .icns files into target_dir from a package archive file.
/// Returns the result code of the pax extract operation.
func extractAppBitsFromPkgArchive(_ archivePath: String, exportDir: String) -> Int {
    if !pathIsRegularFile(archivePath) {
        return -999
    }
    if !pathIsDirectory(exportDir) {
        return -998
    }
    let filemanager = FileManager.default
    let originalWorkingDir = filemanager.currentDirectoryPath
    filemanager.changeCurrentDirectoryPath(exportDir)
    var result = runCLI(
        "/bin/pax",
        arguments: ["-rzf",
                    archivePath,
                    "*.app/Contents/Info.plist",
                    "*.app/Contents/Resources/*.icns"]
    )
    if result.exitcode != 0, pathExists("/usr/bin/aa") {
        // pax failed. Maybe this Payload is an Apple Archive
        result = runCLI(
            "/usr/bin/aa",
            arguments: ["extract",
                        "-i", archivePath,
                        "-include-regex", "\\.app/Contents/Info.plist",
                        "-include-regex", "\\.app/Contents/Resources/.*\\.icns",
                        "-d", "."]
        )
    }
    filemanager.changeCurrentDirectoryPath(originalWorkingDir)
    return result.exitcode
}

/// Extracts application icons from a flat package.
/// Returns a list of paths to icns files.
func extractAppIconsFromFlatPkg(_ pkgPath: String) -> [String] {
    let result = runCLI("/usr/sbin/pkgutil", arguments: ["--bom", pkgPath])
    if result.exitcode != 0 {
        printStderr("Could not get bom files from \(pkgPath): \(result.error)")
        return [String]()
    }
    let bomFilePaths = result.output.components(separatedBy: .newlines)
    var pkgDict = [String: [String]]()
    for bomFile in bomFilePaths {
        // bomfile path is of the form:
        // /tmp/Foo.pkg.boms.2Rxa1z/FooComponent.pkg/Bom
        let tempPath = (bomFile as NSString).deletingLastPathComponent
        var pkgName = (tempPath as NSString).lastPathComponent
        if !pkgName.hasSuffix(".pkg") {
            // pkgPath is not a distribution-style pkg; we're working with
            // a component package
            pkgName = ""
        }
        let result = runCLI("/usr/bin/lsbom", arguments: ["-s", bomFile])
        if result.exitcode != 0 {
            printStderr("Could not get contents of bom: \(bomFile): \(result.error)")
            continue
        }
        let outputLines = result.output.components(separatedBy: .newlines)
        let infoPlistLines = outputLines.filter {
            $0.hasSuffix(".app/Contents/Info.plist")
        }.map {
            ($0 as NSString).standardizingPath
        }
        if !infoPlistLines.isEmpty {
            pkgDict[pkgName] = infoPlistLines
        }
    }
    if pkgDict.isEmpty {
        return [String]()
    }
    var iconPaths = [String]()
    if let pkgTmp = TempDir.shared.makeTempDir(),
       let exportTmp = TempDir.shared.makeTempDir()
    {
        defer {
            try? FileManager.default.removeItem(atPath: pkgTmp)
        }
        let expandedPkgPath = (pkgTmp as NSString).appendingPathComponent("pkg")
        let expandResult = runCLI(
            "/usr/sbin/pkgutil", arguments: ["--expand", pkgPath, expandedPkgPath]
        )
        if expandResult.exitcode == 0 {
            for pkg in pkgDict.keys {
                let archivePath = (expandedPkgPath as NSString).appendingPathComponent("\(pkg)/Payload")
                let err = extractAppBitsFromPkgArchive(archivePath, exportDir: exportTmp)
                if err == 0 {
                    if let infoPaths = pkgDict[pkg] {
                        for infoPath in infoPaths {
                            let fullPath = (exportTmp as NSString).appendingPathComponent(infoPath)
                            let appPath = ((fullPath as NSString).deletingLastPathComponent as NSString).deletingLastPathComponent
                            if let iconPath = findIconForApp(appPath) {
                                iconPaths.append(iconPath)
                            }
                        }
                    }
                } else {
                    printStderr("pax could not read files from \(archivePath)")
                    return iconPaths
                }
            }
        } else {
            printStderr("Could not expand \(pkgPath): \(expandResult.error)")
        }
    }
    return iconPaths
}

/// Returns a list of paths to application Info.plists
func getAppInfoPathsFromBOM(_ bomFile: String) -> [String] {
    var paths = [String]()
    if pathIsRegularFile(bomFile) {
        let result = runCLI("/usr/bin/lsbom", arguments: ["-s", bomFile])
        if result.exitcode == 0 {
            let lines = result.output.components(separatedBy: .newlines)
            paths = lines.filter {
                $0.hasSuffix(".app/Contents/Info.plist")
            }
        }
    }
    return paths
}

/// Returns a dict with pkg paths as keys and filename lists as values
func findInfoPlistPathsInBundlePkg(_ pkgPath: String) -> [String: [String]] {
    var pkgDict = [String: [String]]()
    let bomFile = (pkgPath as NSString).appendingPathComponent("Contents/Archive.bom")
    if pathIsRegularFile(bomFile) {
        // single (simple) package
        let infoPaths = getAppInfoPathsFromBOM(bomFile)
        if !infoPaths.isEmpty {
            pkgDict[pkgPath] = infoPaths
        }
    } else {
        // maybe meta or dist package; look for component pkgs
        var pkgs = [String]()
        let pkgContentsDir = (pkgPath as NSString).appendingPathComponent("Contents")
        if pathIsDirectory(pkgContentsDir) {
            let filemanager = FileManager.default
            let dirEnum = filemanager.enumerator(atPath: pkgContentsDir)
            while let file = dirEnum?.nextObject() as? String {
                if file.hasSuffix(".pkg"), file.hasSuffix(".mpkg") {
                    let fullPath = (pkgContentsDir as NSString).appendingPathComponent(file)
                    pkgs.append(fullPath)
                }
            }
        }
        for pkg in pkgs {
            // Inception time
            let anotherPkgDict = findInfoPlistPathsInBundlePkg(pkg)
            pkgDict.merge(anotherPkgDict) { _, second in second }
        }
    }
    return pkgDict
}

/// Returns a list of paths for application icons found inside the bundle pkg at pkg_path
func extractAppIconsFromBundlePkg(_ pkgPath: String) -> [String] {
    var iconPaths = [String]()
    let pkgDict = findInfoPlistPathsInBundlePkg(pkgPath)
    if let exportTmp = TempDir.shared.makeTempDir() {
        for pkg in pkgDict.keys {
            let archivePath = (pkg as NSString).appendingPathComponent("Contents/Archive.pax.gz")
            if pathIsRegularFile(archivePath) {
                let err = extractAppBitsFromPkgArchive(archivePath, exportDir: exportTmp)
                if err == 0,
                   let infoPaths = pkgDict[pkg]
                {
                    for infoPath in infoPaths {
                        let fullPath = ((exportTmp as NSString).appendingPathComponent(infoPath) as NSString).standardizingPath
                        let appPath = ((fullPath as NSString).deletingLastPathComponent as NSString).deletingLastPathComponent
                        if let iconPath = findIconForApp(appPath) {
                            iconPaths.append(iconPath)
                        }
                    }
                }
            }
        }
    }
    return iconPaths
}
